pkg/kubelet/cm/dra/plugin/plugin.go

```diff
@@ -18,94 +18,140 @@ package plugin
+	kubeClient kubernetes.Interface
+	restConfig *rest.Config
+	getNode    func() (*v1.Node, error)
 }
 
+var _ cache.PluginHandler = &RegistrationHandler{}
+
 // NewPluginHandler returns new registration handler.
 //
 // Must only be called once per process because it manages global state.
 // If a kubeClient is provided, then it synchronizes ResourceSlices
 // with the resource information provided by plugins.
-func NewRegistrationHandler(kubeClient kubernetes.Interface, getNode func() (*v1.Node, error)) *RegistrationHandler {
-	handler := &RegistrationHandler{}
+func NewRegistrationHandler(kubeClient kubernetes.Interface, restConfig *rest.Config, getNode func() (*v1.Node, error)) *RegistrationHandler {
+	handler := &RegistrationHandler{
+		// The context and thus logger should come from the caller.
+		ctx:        klog.NewContext(context.TODO(), klog.LoggerWithName(klog.TODO(), "DRA registration handler")),
+		kubeClient: kubeClient,
+		restConfig: restConfig,
+		getNode:    getNode,
+	}
 
-	// If kubelet ever gets an API for stopping registration handlers, then
-	// that would need to be hooked up with stopping the controller.
-	handler.controller = startNodeResourcesController(context.TODO(), kubeClient, getNode)
+	// When kubelet starts up, no DRA driver has registered yet. None of
+	// the drivers are usable until they come back, which might not happen
+	// at all. Therefore it is better to not advertise any local resources
+	// because pods could get stuck on the node waiting for the driver
+	// to start up.
+	//
+	// This has to run in the background.
+	go handler.wipeResourceSlices("")
 
 	return handler
 }
 
+// wipeResourceSlices deletes ResourceSlices of the node, optionally just for a specific driver.
+func (h *RegistrationHandler) wipeResourceSlices(pluginName string) {
+	if h.restConfig == nil {
+		return
+	}
+	ctx := h.ctx
+	logger := klog.FromContext(ctx)
+	backoff := wait.Backoff{
+		Duration: time.Second,
+		Factor:   2,
+		Jitter:   0.2,
+		Cap:      5 * time.Minute,
+		Steps:    100,
+	}
+	discoveryCache := memory.NewMemCacheClient(h.kubeClient.Discovery())
+	restMapper := restmapper.NewDeferredDiscoveryRESTMapper(discoveryCache)
+	gk := resourceapi.SchemeGroupVersion.WithKind("ResourceSlice").GroupKind()
+	dynClient, err := dynamic.NewForConfig(h.restConfig)
+	if err != nil {
+		logger.Error(err, "Creating dynamic client for REST config failed, not deleting ResourceSlices")
+		return
+	}
+
+	// Error logging is done inside the loop. Context cancellation doesn't get logged.
+	_ = wait.ExponentialBackoffWithContext(ctx, backoff, func(ctx context.Context) (bool, error) {
+		node, err := h.getNode()
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		if err != nil {
+			logger.Error(err, "Unexpected error checking for node")
+		}
+		fieldSelector := fields.Set{"nodeName": node.Name}
+		if pluginName != "" {
+			fieldSelector["driverName"] = pluginName
+		}
+
+		mapping, err := restMapper.RESTMapping(gk)
```

Comment:
why are we using discovery here? the kubelet should assume the apiserver is as new as it is and is serving the version(s) the kubelet knows about

Reply:
I had the same thought at some point (it's in the KEP update) and I thought I had changed the code, but obviously not. Will fix.